BOLT01: TLV proposal (#607)
lightningnetwork/lightning-rfc: 01-messaging.md

  * [Connection Handling and Multiplexing](#connection-handling-and-multiplexing)
  * [Lightning Message Format](#lightning-message-format)

>   * [Wire Type-Length-Value Format](#wire-type-length-value-format)

Reviewer: nit: maybe no need to specify `wire` now that we agree to use this everywhere. Simply `Type-Length-Value Format`?

wasteful: alignment may be achieved by decrypting the message into
a buffer with 6-bytes of pre-padding.

## Wire Type-Length-Value Format

Throughout the protocol, we use a TLV (Type-Length-Value) format to allow for
the backwards-compatible addition of new fields to existing message types. This
format describes the TLV encoding specific to wire messages.

A `tlv_record` represents a single field, encoded in the form:

* [`1`: `type`]
* [`2`: `length`]
* [`length`: `value`]

A `tlv_stream` is a series of (possibly zero) `tlv_record`s, represented as the
concatenation of the encoded `tlv_record`s. When used to extend existing
messages, a `tlv_stream` is typically placed after all currently defined fields.

The `type` is a single-byte identifier for the `tlv_record` and determines how
the contents of `value` should be decoded. The `type` 255 is reserved for use as
a sentinel value indicating the end of the `tlv_stream`; the length of `type`
255 is always zero.

> 

Reviewer 2: I have question that also applies to @rustyrussell  PR: what is the actual use case for the sentinel/end-of-stream type ?
Since readers must know how to parse T and L (and then skip over L bytes if they don't need or understand T)  I don't really see why it is needed ?

Author: you're correct that it's not strictly needed. however, one application is that you can concatenate two `tlv_stream`s with different namespaces. assuming the receiver is asserting the ordering of `type`s, it would otherwise be impossible to distinguish between the stream being not-canonical and the second one starting.

it's also useful in signaling to the receiver that all intended fields were serialized, and that nothing got cutoff/dropped. it could make sense to always include the sentinel value so that this is implicit, but i think maybe that could be done on a case-by-base basis depending on the criticality of the message and its contents

i'm sure @rustyrussell has some more ideas as well :)

Author: after further discussion, seems we no longer need sentinel values :) removed in 252ecea1d262c9959159358b7fed38f784bcd96a

   - if `type` is unknown:
     - if the `tlv_stream` is not covered by a signature:
       - MUST discard the next `length` bytes.
     - otherwise:
       - MUST retain a `tlv_record` for `type` using the next `length` bytes as
         its `value`.
 - if the `type`s in a `tlv_stream` are not monotonically-increasing:
   - MUST fail to parse the `tlv_stream`.
 - if the `length` of `type` 255 is not 0:
   - MUST fail to parse the `tlv_stream`.

### Rationale

The primary advantage in using TLV is that a reader is able to ignore new fields
that it does not understand, since each field carries the exact size of the
encoded element. Without TLV, even if a node does not wish to use a particular
field, the node is forced to add parsing logic for that field in order to
determine the offset of any fields that follow.

The sentinel value is chosen as 255 since it maintains that all records are
sorted in ascending order, and implicitly skips over all possible `type`
identifiers. Including a sentinel value allows, for example, concatenating
`tlv_stream`s with different namespaces.

The monotonicity constraint ensures that all `type`s are unique and can appear

> at most once.

Reviewer: Is there a reason why you want to restrict types to appear at most once in a tlv stream? I think it could be useful and more flexible to let them appear multiple times.

Author: replied in main body where there was ample space 😬

Reviewer: I think it could be useful to add a sentence here to explain that if multiple occurrences of the same type is needed, it should rather be done at the type definition level (so we won't need to re-discuss this when newcomers read the spec and wonder why this limitation is here).

Author: added in 2583ae291eaa672f12da4a35f057ee1f48f71871

 - SHOULD NOT use the varbytes encoding for `value`s containing byte arrays.

The receiving node:
 - if encounters EOF while parsing a `type`:
   - MUST stop parsing the `tlv_stream`.
 - if encounters EOF while parsing a `length` or `value`:
   - MUST fail to parse to `tlv_stream`.
 - if `type` and `length` parse:
   - if `type` is known:
     - MUST decode the next `length` bytes using the known encoding for `type`.
   - otherwise, if `type` is unknown:
       - MUST discard the next `length` bytes.
 - if the `type`s in a `tlv_stream` are not monotonically-increasing:
   - MUST fail to parse the `tlv_stream`.

### Rationale

The primary advantage in using TLV is that a reader is able to ignore new fields
that it does not understand, since each field carries the exact size of the
encoded element. Without TLV, even if a node does not wish to use a particular
field, the node is forced to add parsing logic for that field in order to
determine the offset of any fields that follow.

The monotonicity constraint ensures that all `type`s are unique and can appear
at most once.

The use of a varint for `length` permits a space savings over a fixed 16-bit

> `length` for `value`s whose encode is less than 253 bytes. When `tlv_stream` is

Author: good catch, fixed in 2583ae291eaa672f12da4a35f057ee1f48f71871

> used in EOB multi-frame payload, this can potentially save payloads from

Author: You're right, I don't think it is. I ended up removing it and just referring to them as multi-frame payloads, since this better matches the language in #593 

2583ae291eaa672f12da4a35f057ee1f48f71871

wasteful: alignment may be achieved by decrypting the message into
a buffer with 6-bytes of pre-padding.

## Type-Length-Value Format

> Throughout the protocol, we use a TLV (Type-Length-Value) format to allow for

Reviewer 3: nit: specs should not use `we`; (who is we?)

```suggestion
Throughout the protocol, a TLV (Type-Length-Value) format is used to allow for
```

the backwards-compatible addition of new fields to existing message types.

A `tlv_record` represents a single field, encoded in the form:

* [`varint`: `type`]
* [`varint`: `length`]
* [`length`: `value`]

A `tlv_stream` is a series of (possibly zero) `tlv_record`s, represented as the
concatenation of the encoded `tlv_record`s. When used to extend existing
messages, a `tlv_stream` is typically placed after all currently defined fields.

The `type` is a varint encoded using the bitcoin CompactSize format. It
functions as a message-specific, 64-bit identifier for the `tlv_record`
determining how the contents of `value` should be decoded.

The `length` is a varint encoded using the bitcoin CompactSize format
signaling the size of `value` in bytes.

The `value` depends entirely on the `type`, and should be encoded or decoded
according to the message-specific format determined by `type`.

### Requirements

The sending node:
 - MUST order `tlv_record`s in a `tlv_stream` by monotonically-increasing `type`.
 - MUST minimally encode `type` and `length`.

>  - SHOULD NOT use the varbytes encoding for `value`s containing byte arrays.

Reviewer 3: this is the first time we're using `varbytes` in the spec, should we provide a definition (or at least a link to a definition)?

Author: ended up removing the mention of varbytes in favor of rusty's language from the irc meeting, which says not to use redundant, variable-length encodings

The receiving node:
 - if zero bytes remain before parsing a `type`:
   - MUST stop parsing the `tlv_stream`.

>  - if a `type` or `length` is not minimally encoded:

Reviewer 3: clarification: is `minimally encoded` referring to their being a varint type, i.e. you should encode '0x01' as '0x01', not '0xfe00000001'?

Author: @niftynei correct, i would assume most impls of compact size already check this, tho i'm not positive

Reviewer 4: `btcd`'s wire package includes checks to ensure encoded integers are canonical: https://github.com/btcsuite/btcd/blob/master/wire/common.go#L491

   - MUST fail to parse the `tlv_stream`.
 - if decoded `type`s are not monotonically-increasing:
   - MUST fail to parse the `tlv_stream`.
 - if `type` is known:
   - MUST decode the next `length` bytes using the known encoding for `type`.
 - otherwise, if `type` is unknown:
   - if `type` is even:

>      - MUST fail to parse the `tlv_strem`.

Reviewer 5: ```suggestion
     - MUST fail to parse the `tlv_stream`.
```

serialized within a single `tlv_record`. The uniqueness constraint, among other
things, enables the following optimizations:
 - canonical ordering is defined independent of the encoded `value`s.
 - canonical ordering can be known at compile-time, rather that being determined
   dynamically at the time of encoding.
 - verifying canonical ordering requires less state and is less-expensive.
 - variable-size fields can reserve their expected size up front, rather than
   appending elements sequentially and incurring double-and-copy overhead.

The use of a varint for `type` and `length` permits a space savings for small
`type`s or short `value`s. This potentially leaves more space for application
data over the wire or in an onion payload.

All `type`s must appear in increasing order to create a canonical encoding of
the underlying `tlv_record`s. This is crucial when computing signatures over a
`tlv_stream`, as it ensures verifiers will be able to recompute the same message
digest as the signer. Note that the canonical ordering over the set of fields
can be enforced even if the verifier does not understand what the fields
contain.

> We recommend that writers not use the varbytes encoding for byte arrays since it

Reviewer 3: Small nit that `We` is kind of weird tone for a spec doc.

```suggestion
Writers should not use the varbytes encoding for byte arrays since it
```

Author: heh, old habits die hard 😅fixed

wasteful: alignment may be achieved by decrypting the message into
a buffer with 6-bytes of pre-padding.

## Wire Type-Length-Value Format

Throughout the protocol, we use a TLV (Type-Length-Value) format to allow for
the backwards-compatible addition of new fields to existing message types. This
format describes the TLV encoding specific to wire messages.

A `tlv_record` represents a single field, encoded in the form:

* [`varint`: `type`]
* [`varint`: `length`]
* [`length`: `value`]

A `tlv_stream` is a series of (possibly zero) `tlv_record`s, represented as the
concatenation of the encoded `tlv_record`s. When used to extend existing
messages, a `tlv_stream` is typically placed after all currently defined fields.

The `type` is a varint encoded using the bitcoin CompactSize format. It
functions as a message-specific, 64-bit identifier for the `tlv_record`
determining how the contents of `value` should be decoded.

The `length` is a varint encoded using the bitcoin CompactSize format
signaling the size of `value` in bytes.

> The `value` depends entirely on the `type`. Each `tlv_stream` has its own 64-bit

Reviewer: Could you detail that a bit?
Does that mean that a `tlv_stream` MUST have the following format?

- [`8`:`namespace`]
- [`varint`:`length`]
- TLV records

It seems incompatible with the current wire lightning message format that use a 2-byte `type` (unless we then encode the namespace in the `payload` part).

I thought the `namespace` would be detected somewhat implicitly. For example in onion records, the `namespace` would be the onion version, and in wire messages it would be the 2-byte top-level `type`.

Author: Yeah the namespace isn’t encoded, it’s implied from the context (wire message, onion payload+realm byte, etc).

Which parts do you think need clarification?

Reviewer: Ok then great, I didn't understand why you explicitly said that it uses a 64-bit namespace, maybe that part should be removed?

Author: Types are now 64-bit unsigned integers because of the use of CompactSize, hence each message has a 64-bit namespace. I added it to clarify the range of possible values a `type` can assume, but maybe there’s a better way to say that?

Reviewer: Gotcha I slightly misunderstood the sentence, I thought it meant a tlv stream namespace needed to be identified by a 64-bit identifier. With your explanation the sentence makes more sense.
I'll try to think of different wordings, but nothing better comes to mind right now...

Author: removed the mention of 64-bit namespace from `value`, since it seemed redundant after already being addressed in the description of `type`

the backwards-compatible addition of new fields to existing message types.

A `tlv_record` represents a single field, encoded in the form:

* [`varint`: `type`]
* [`varint`: `length`]
* [`length`: `value`]

A `tlv_stream` is a series of (possibly zero) `tlv_record`s, represented as the
concatenation of the encoded `tlv_record`s. When used to extend existing
messages, a `tlv_stream` is typically placed after all currently defined fields.

The `type` is a varint encoded using the bitcoin CompactSize format. It
functions as a message-specific, 64-bit identifier for the `tlv_record`
determining how the contents of `value` should be decoded.

The `length` is a varint encoded using the bitcoin CompactSize format
signaling the size of `value` in bytes.

The `value` depends entirely on the `type`, and should be encoded or decoded
according to the message-specific format determined by `type`.

### Requirements

The sending node:
 - MUST order `tlv_record`s in a `tlv_stream` by monotonically-increasing `type`.
 - MUST minimally encode `type` and `length`.
 - SHOULD NOT use redundant, variable-length encodings in a `tlv_record`.

> The receiving node:

Reviewer 4: What about rejecting redundant types in the stream?

Reviewer 4: I guess this is covered by the "strictly increasing" requirement.
